pkg/ipmi/ipmi.go

```diff
@@ -173,3 +191,54 @@ func (i *IPMI) ShutoffWatchdog() error {
 
 	return nil
 }
+
+func (i *IPMI) setsysinfo(data *setSystemInfoReq) error {
+	req := &req{}
+	req.msg.cmd = _SET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	req.msg.dataLen = 18 // size of setSystemInfoReq
+	req.msg.data = unsafe.Pointer(data)
+
+	if _, err := i.sendrecv(req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetSystemFWVersion sets the provided system firmware version to BMC via IPMI.
+func (i *IPMI) SetSystemFWVersion(version string) error {
+	len := len(version)
+
+	if len == 0 {
+		return fmt.Errorf("Version length is 0")
+	} else if len > strlenMax {
+		len = strlenMax
+	}
+
+	var data setSystemInfoReq
+	var copied int
+	var index int
+	data.paramSelector = _SYSTEM_FW_VERSION
+	data.setSelector = 0
+	for len > index {
+		if data.setSelector == 0 { // the fisrt block of string data
+			data.strData[0] = _ASCII
+			data.strData[1] = byte(len)
+			copied = copy(data.strData[2:], version)
```

Comment:
* How about defining a `strncpy` function that copies the data and fills the rest with zeros to avoid the reset in line 238?
* Is the copied variable needed? I think you can do `index += ...` and remove the `copied` variable.
* This will always return 14. Should we do index += 14 for clarity?

Reply:
I added a strcpyPadded function for it, because in our case it's not exactly the same as C strncpy.